source/common/http/http2/codec_impl.cc

```diff
@@ -506,24 +506,59 @@ void ConnectionImpl::sendPendingFrames() {
   }
 }
 
-void ConnectionImpl::sendSettings(uint64_t codec_options) {
-  std::vector<nghttp2_settings_entry> iv = {
-      {NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, MAX_CONCURRENT_STREAMS},
-      {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, DEFAULT_WINDOW_SIZE}};
+void ConnectionImpl::sendSettings(const Http2Settings& http2_settings) {
+  ASSERT(http2_settings.hpack_table_size_ <= Http2Settings::MAX_HPACK_TABLE_SIZE);
+  ASSERT(Http2Settings::MIN_MAX_CONCURRENT_STREAMS <= http2_settings.max_concurrent_streams_ &&
+         http2_settings.max_concurrent_streams_ <= Http2Settings::MAX_MAX_CONCURRENT_STREAMS);
+  ASSERT(
+      Http2Settings::MIN_INITIAL_STREAM_WINDOW_SIZE <= http2_settings.initial_stream_window_size_ &&
+      http2_settings.initial_stream_window_size_ <= Http2Settings::MAX_INITIAL_STREAM_WINDOW_SIZE);
+  ASSERT(Http2Settings::MIN_INITIAL_CONNECTION_WINDOW_SIZE <=
+             http2_settings.initial_connection_window_size_ &&
+         http2_settings.initial_connection_window_size_ <=
+             Http2Settings::MAX_INITIAL_CONNECTION_WINDOW_SIZE);
+
+  std::vector<nghttp2_settings_entry> iv;
```

Comment:
For posterity, for perf we should switch this to a fixed size array with a pre-defined max size. Added to https://github.com/lyft/envoy/issues/210.